### PR TITLE
Boto ec2 eni updates

### PR DESCRIPTION
**Pull request by lomeroe**
Modify associate_eip_address function in boto_ec2 module to allow associating an EIP to an ENI that is not yet attached to an instance (remove requirement for instance_id or instance_name to have a value)

Modify boto_ec2 state to allocate/associate an EIP to an ENI on eni_present
Modify boto_ec2 state to allow creation of DNS A records (similar to cname attribute in boto_elb)
Modify boto_ec2 state to allow release of EIP on eni_absent

@ryan-lane : FYI

### CODE REVIEW

File: salt/states/boto_ec2.py [hunk 1]
```diff
@@ -172,7 +172,9 @@ def eni_present(
         region=None,
         key=None,
         keyid=None,
-        profile=None):
+        profile=None,
+        allocate_eip=False,
+        arecords=None):
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Can you place these above region?

File: salt/states/boto_ec2.py [hunk 2]
```diff
@@ -278,6 +292,79 @@ def eni_present(
     ret['comment'] = ' '.join([ret['comment'], _ret['comment']])
     if not _ret['result']:
         ret['result'] = _ret['result']
+        return ret
+    if allocate_eip:
+        if 'allocationId' not in r['result']:
+            eip_alloc = __salt__['boto_ec2.allocate_eip_address'](domain=None,
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Needs to support test mode.

File: salt/states/boto_ec2.py [hunk 3]
```diff
@@ -278,6 +292,79 @@ def eni_present(
...
+                                                                region=region,
+                                                                key=key,
+                                                                keyid=keyid,
+                                                                profile=profile)
+            if eip_alloc:
+                _ret = __salt__['boto_ec2.associate_eip_address'](instance_id=None,
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Needs to support test mode.

File: salt/states/boto_ec2.py [hunk 4]
```diff
@@ -278,6 +292,79 @@ def eni_present(
...
+                                                                  allocation_id=eip_alloc['allocation_id'],
+                                                                  network_interface_id=r['result']['id'],
+                                                                  private_ip_address=None,
+                                                                  allow_reassociation=False,
+                                                                  region=region,
+                                                                  key=key,
+                                                                  keyid=keyid,
+                                                                  profile=profile)
+                if not _ret:
+                    _ret = __salt__['boto_ec2.release_eip_address'](public_ip=None,
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Needs to support test mode.

File: salt/states/boto_ec2.py [hunk 5]
```diff
@@ -412,6 +505,20 @@ def eni_absent(
             return ret
         ret['comment'] = 'Deleted ENI {0}'.format(name)
         ret['changes']['id'] = None
+        if release_eip and 'allocationId' in r['result']:
+            _ret = __salt__['boto_ec2.release_eip_address'](public_ip=None,
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Needs to support test mode.

**Comment by lomeroe** [comment]
@ryan-lane thanks for the quick code check, I added test support for arecords/allocate_eip in the eni_present.

In eni_absent, there is a check/statement in the "if __opts__['test'] (line 493).  Any test code in the later block would never get executed b/c of the return in that if block.  Or am I missing something?

File: salt/states/boto_ec2.py [hunk 6]
```diff
@@ -386,6 +492,8 @@ def eni_absent(
     else:
         if __opts__['test']:
             ret['comment'] = 'ENI is set to be deleted.'
+            if release_eip and 'allocationId' in r['result']:
```
Comment (lomeroe) [inline salt/states/boto_ec2.py]: @ryan-lane wouldn't this be the only check needed for a test=True state run?  Or am I missing something?

Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Oh. yes. I'll delete the other note.

File: salt/states/boto_ec2.py [hunk 7]
```diff
@@ -278,6 +296,90 @@ def eni_present(
...
+                    ret['comment'] = ' '.join([ret['comment'], 'Failed to allocate an EIP address'])
+                    return ret
+        else:
+            ret['comment'] = ' '.join([ret['comment'], 'An EIP is already allocated/assocaited to the ENI'])
+    if arecords:
+        for arecord in arecords:
+            if 'name' not in arecord:
+                msg = 'The arecord must contain a "name" property.'
+                raise SaltInvocationError(msg)
+            if __opts__['test']:
```
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: I think it's actually ok to not have test mode here. The state module you're calling has test mode and will handle this correctly.

**Comment by cachedout** [comment]
Go Go Jenkins!

File: salt/states/boto_ec2.py [hunk 8]
```diff
@@ -169,6 +169,8 @@ def eni_present(
         description=None,
         groups=None,
         source_dest_check=True,
+        allocate_eip=False,
```
Comment (cachedout) [inline salt/states/boto_ec2.py]: We generally prefer to add new kwargs to the end of the list, since python allows kwargs to be passed positionally.

**Comment by cachedout** [comment]
Hi @lomeroe. Just one bit of feedback from me that I put in-line. Please have a look and then we'll get this merged. Looks great!

**Comment by lomeroe** [comment]
@cachedout I don't mind moving the arguments back to the bottom, @ryan-lane requested moving them above region (first line note above)

you guys want to arm wrestle for it? :)

File: salt/states/boto_ec2.py (hunk 8 above)
Comment (ryan-lane) [inline salt/states/boto_ec2.py]: Region, key, keyid, and profile are never expected to be passed positionally. My preference is for new args to be before region.

Comment (cachedout) [inline salt/states/boto_ec2.py]: You know this code far better than I do, so I happily defer to you. Thanks, @ryan-lane :]

**Comment by cachedout** [comment]
@lomeroe He knows this particular code much better than I do. I trust his opinon. :] I'll go ahead and merge this.

**Pull request merged**